### PR TITLE
[Clang][AMDGPU] Add builtins for instrinsic `llvm.amdgcn.raw.ptr.buffer.load`

### PR DESCRIPTION
**Pull request by shiltian**

### CODE REVIEW

**Comment by shiltian** [comment]
* **#99258** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/99258?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -19177,6 +19177,45 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
...
+      break;
+    case AMDGPU::BI__builtin_amdgcn_raw_buffer_load_b16:
+      RetTy = Int16Ty;
+      break;
+    case AMDGPU::BI__builtin_amdgcn_raw_buffer_load_b32:
+      RetTy = Int32Ty;
+      break;
+    case AMDGPU::BI__builtin_amdgcn_raw_buffer_load_b64:
+      RetTy =
+          llvm::VectorType::get(Int32Ty, /*NumElements=*/2, /*Scalable=*/false);
```
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:19199]: Using FixedVectorType would avoid the scalable parameter and line wrap

File: clang/include/clang/Basic/BuiltinsAMDGPU.def [hunk 2]
```diff
@@ -155,6 +155,12 @@ BUILTIN(__builtin_amdgcn_raw_buffer_store_b32, "viQbiiIi", "n")
 BUILTIN(__builtin_amdgcn_raw_buffer_store_b64, "vV2iQbiiIi", "n")
 BUILTIN(__builtin_amdgcn_raw_buffer_store_b96, "vV3iQbiiIi", "n")
 BUILTIN(__builtin_amdgcn_raw_buffer_store_b128, "vV4iQbiiIi", "n")
+BUILTIN(__builtin_amdgcn_raw_buffer_load_b8, "cQbiiIi", "n")
+BUILTIN(__builtin_amdgcn_raw_buffer_load_b16, "sQbiiIi", "n")
+BUILTIN(__builtin_amdgcn_raw_buffer_load_b32, "iQbiiIi", "n")
+BUILTIN(__builtin_amdgcn_raw_buffer_load_b64, "V2iQbiiIi", "n")
+BUILTIN(__builtin_amdgcn_raw_buffer_load_b96, "V3iQbiiIi", "n")
```
Comment (arsenm) [inline clang/include/clang/Basic/BuiltinsAMDGPU.def:162]: Technically this one depends on a subtarget feature for 96 bit load/store but I don't see one that applies yet

Comment (shiltian) [reply clang/include/clang/Basic/BuiltinsAMDGPU.def:162]: We can do a separate patch to add the sub target feature to change both of them.

Comment (arsenm) [inline clang/include/clang/Basic/BuiltinsAMDGPU.def:162]: These should be returning unsigned results? The stores also should have been unsigned

Comment (arsenm) [reply clang/include/clang/Basic/BuiltinsAMDGPU.def:162]: For the 8/16-bit cases, we could also return the 32-bit result they really return and insert the sext/zext we fold into the instruction during codegen. Not sure if that would be better

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 3]
```diff
@@ -19177,6 +19177,42 @@ Value *CodeGenFunction::EmitAMDGPUBuiltinExpr(unsigned BuiltinID,
...
+      break;
+    case AMDGPU::BI__builtin_amdgcn_raw_buffer_load_b128:
+      RetTy = llvm::FixedVectorType::get(Int32Ty, /*NumElements=*/4);
+      break;
+    }
+    SmallVector<Value *, 4> Args;
+    Args.push_back(EmitScalarExpr(E->getArg(0)));
+    Args.push_back(EmitScalarExpr(E->getArg(1)));
+    Args.push_back(EmitScalarExpr(E->getArg(2)));
+    Args.push_back(EmitScalarExpr(E->getArg(3)));
```
Comment (arsenm) [inline clang/lib/CodeGen/CGBuiltin.cpp:19211]: Can do this in initializer list

File: clang/include/clang/Basic/BuiltinsAMDGPU.def (hunk 2 above)
Comment (shiltian) [reply clang/include/clang/Basic/BuiltinsAMDGPU.def:162]: As long as the size matches, it doesn't matter too much whether it is signed or unsigned I think? I can make them unsgined if necessary.

Comment (arsenm) [reply clang/include/clang/Basic/BuiltinsAMDGPU.def]: It matters for the usage context

**Review by arsenm — APPROVED** [review]
probably should change the stores to unsigned to match in a separate patch

**Comment by shiltian** [comment]
> probably should change the stores to unsigned to match in a separate patch

Doing it right now.

**Pull request merged**